Assignment 7 - ANN and SVM

Tyler Gardner

10/31/2021

# Task 1

## Code Chunk 1 - Set up, Data Import, and Preparation

In code chunk 1 we set up the Colab notebook by importing the packages that are needed to run our code and analyze the data. The google drive is mounted to import and read the data. The data is then looked over and analyzed. The target variable is popped from the data set and encoded. The data set is then encoded as well so we can explore and analyze the data set.

In [1]:
# A. Data Import

# Import Packages Needed
import pandas as pd
import numpy as np
import sklearn

from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix,\
 recall_score, precision_score, f1_score, accuracy_score, make_scorer,\
  precision_recall_fscore_support
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

import warnings
warnings.filterwarnings('ignore')

In [2]:
# mount the google drive location to read the dataset
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Import data using pd.read_csv()
CD_additional_modified_v2_raw = pd.read_csv("/content/drive/MyDrive/data_sets/CD_additional_modified-v2.csv")
CD_additional_modified_v2 = CD_additional_modified_v2_raw.copy()

In [4]:
# B. Show the overall structure and summary of the data frame using info(), describe(), head()

CD_additional_modified_v2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4117 entries, 0 to 4116
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      4117 non-null   int64  
 1   age             4117 non-null   int64  
 2   job             4117 non-null   object 
 3   marital         4117 non-null   object 
 4   education       4117 non-null   object 
 5   default         4117 non-null   object 
 6   housing         4117 non-null   object 
 7   loan            4117 non-null   object 
 8   contact         4117 non-null   object 
 9   month           4117 non-null   object 
 10  day_of_week     4117 non-null   object 
 11  duration        4117 non-null   int64  
 12  campaign        4117 non-null   int64  
 13  pdays           4117 non-null   int64  
 14  previous        4117 non-null   int64  
 15  poutcome        4117 non-null   object 
 16  emp_var_rate    4117 non-null   float64
 17  cons_price_idx  4117 non-null   f

In [5]:
CD_additional_modified_v2.describe()

Unnamed: 0          age  ...    euribor3m  nr_employed
count  4117.000000  4117.000000  ...  4117.000000  4117.000000
mean   2058.193102    40.115375  ...     3.621904  5166.496502
std    1188.913501    10.314847  ...     1.733448    73.670942
min       0.000000    18.000000  ...     0.635000  4963.600000
25%    1029.000000    32.000000  ...     1.334000  5099.100000
50%    2058.000000    38.000000  ...     4.857000  5191.000000
75%    3087.000000    47.000000  ...     4.961000  5228.100000
max    4118.000000    88.000000  ...     5.045000  5228.100000

[8 rows x 11 columns]

In [6]:
CD_additional_modified_v2.head()

Unnamed: 0  age          job  ... euribor3m nr_employed   y
0           0   30  blue-collar  ...     1.313      5099.1  no
1           1   39     services  ...     4.855      5191.0  no
2           2   25     services  ...     4.962      5228.1  no
3           3   38     services  ...     4.959      5228.1  no
4           4   47       admin.  ...     4.191      5195.8  no

[5 rows x 22 columns]

In [7]:
# C. pop() the target variable into a new variable ie y_target
y_target = CD_additional_modified_v2.pop('y')

In [8]:
# C1. Encode the target variable using the following code: y_target = pd.get_dummies(y_target, drop_first = True)
y_target = pd.get_dummies(y_target,drop_first=True)

In [9]:
# D. Show the overall structure and summary of the data frame using info(), describe(), head()
CD_additional_modified_v2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4117 entries, 0 to 4116
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      4117 non-null   int64  
 1   age             4117 non-null   int64  
 2   job             4117 non-null   object 
 3   marital         4117 non-null   object 
 4   education       4117 non-null   object 
 5   default         4117 non-null   object 
 6   housing         4117 non-null   object 
 7   loan            4117 non-null   object 
 8   contact         4117 non-null   object 
 9   month           4117 non-null   object 
 10  day_of_week     4117 non-null   object 
 11  duration        4117 non-null   int64  
 12  campaign        4117 non-null   int64  
 13  pdays           4117 non-null   int64  
 14  previous        4117 non-null   int64  
 15  poutcome        4117 non-null   object 
 16  emp_var_rate    4117 non-null   float64
 17  cons_price_idx  4117 non-null   f

In [10]:
CD_additional_modified_v2.describe()

Unnamed: 0          age  ...    euribor3m  nr_employed
count  4117.000000  4117.000000  ...  4117.000000  4117.000000
mean   2058.193102    40.115375  ...     3.621904  5166.496502
std    1188.913501    10.314847  ...     1.733448    73.670942
min       0.000000    18.000000  ...     0.635000  4963.600000
25%    1029.000000    32.000000  ...     1.334000  5099.100000
50%    2058.000000    38.000000  ...     4.857000  5191.000000
75%    3087.000000    47.000000  ...     4.961000  5228.100000
max    4118.000000    88.000000  ...     5.045000  5228.100000

[8 rows x 11 columns]

In [11]:
CD_additional_modified_v2.head()

Unnamed: 0  age          job  ... cons_conf_idx euribor3m nr_employed
0           0   30  blue-collar  ...         -46.2     1.313      5099.1
1           1   39     services  ...         -36.4     4.855      5191.0
2           2   25     services  ...         -41.8     4.962      5228.1
3           3   38     services  ...         -41.8     4.959      5228.1
4           4   47       admin.  ...         -42.0     4.191      5195.8

[5 rows x 21 columns]

In [12]:
# E. Encode the dataframe using pd.get_dummies()
CD_additional_modified_v2_enc = pd.get_dummies(CD_additional_modified_v2)

# Task 2

## Code Chunk 2 - ANN Models

In code chunk 2 multiple models MLP models are trained using 5 fold cross validation for each one to evaluate their performance for accuracy, precision, recall and F1. The Cross validation results are shown as a data frame for each model that is trained. Each model is built and cross validated individually. After running the various MLP models a text box explains which model appears to have the best average f1 score by examining the results by visual inspection. 

In [13]:
# 1 Train multiple MLP models and use 5 fold cross validation for each one to evaluate their performance for the following metrics
      # 1 Accuracy, Prcision, Recall, F1. The CV results should be shown as a dataframe for each mdoel as shown in the tutorial code.

In [14]:
# 2. Do NOT use GridSearchCV for this section
    # Each model must be built and cross validated individually
      # A MLP model for each of the following
            # 1 - 1 Hidden Layer with 7 neurons
            # 2 - 2 hidden layers each with 30 neurons
            # 3 - 3 hidden layers, the first has 25 neurons and the second and third have 10 neurons
            # 4 - 2 hidden layers each with 20 neurons
            # 5 - 2 hidden layers each with 10 neurons

In [15]:
# 1 - 1 Hidden Layer with 7 neurons
model_1 = MLPClassifier(hidden_layer_sizes=(7,)).fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_1.hidden_layer_sizes)
print("n_layers_",model_1.n_layers_)

hidden layers sizes (7,)
n_layers_ 3


In [16]:
model_1_cv_results = pd.DataFrame(cross_validate(model_1,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_1_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  0.118211    0.009051       0.900485  ...         0.472296  0.422535  0.483784
1  0.704338    0.006886       0.916262  ...         0.672131  0.481203  0.453039
2  0.949402    0.007127       0.911300  ...         0.681818  0.451128  0.446927
3  0.556031    0.007811       0.908870  ...         0.705882  0.475524  0.536082
4  0.143168    0.007101       0.791009  ...         0.265934  0.245614  0.296569

[5 rows x 10 columns]

In [17]:
# 2 - 2 hidden layers each with 30 neurons
model_2 = MLPClassifier(hidden_layer_sizes=(30,30)).fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_2.hidden_layer_sizes)
print("n_layers_",model_2.n_layers_)

hidden layers sizes (30, 30)
n_layers_ 4


In [18]:
model_2_cv_results = pd.DataFrame(cross_validate(model_2,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_2_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  1.741882    0.008416       0.559466  ...         0.807692  0.176871  0.108527
1  0.686569    0.008230       0.905340  ...         0.636691  0.551724  0.554859
2  1.929290    0.008228       0.891859  ...         0.620690  0.479532  0.552995
3  0.809253    0.007810       0.900365  ...         0.637500  0.378788  0.509151
4  1.418722    0.008399       0.878493  ...         0.623656  0.479167  0.543750

[5 rows x 10 columns]

In [19]:
# 3 - 3 hidden layers, the first has 25 neurons and the second and third have 10 neurons
model_3 = MLPClassifier(hidden_layer_sizes=(25,10,10)).fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_3.hidden_layer_sizes)
print("n_layers_",model_3.n_layers_)

hidden layers sizes (25, 10, 10)
n_layers_ 5


In [20]:
model_3_cv_results = pd.DataFrame(cross_validate(model_3,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_3_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  4.814354    0.007902       0.867718  ...         0.555556  0.384181  0.581242
1  3.242280    0.007692       0.904126  ...         0.604240  0.439716  0.531882
2  2.184715    0.007685       0.910085  ...         0.607903  0.564706  0.579710
3  1.686718    0.007710       0.907655  ...         0.665138  0.449275  0.500864
4  4.876196    0.008240       0.890644  ...         0.000000  0.000000  0.000000

[5 rows x 10 columns]

In [21]:
# 4 - 2 hidden layers each with 20 neurons
model_4 = MLPClassifier(hidden_layer_sizes=(20,20)).fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_4.hidden_layer_sizes)
print("n_layers_",model_4.n_layers_)

hidden layers sizes (20, 20)
n_layers_ 4


In [22]:
model_4_cv_results = pd.DataFrame(cross_validate(model_4,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_4_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  0.449926    0.007740       0.402913  ...         0.701493  0.151724  0.379798
1  0.744052    0.009927       0.832524  ...         0.405868  0.551948  0.563667
2  0.517278    0.010266       0.890644  ...         0.763636  0.042553  0.356688
3  0.985015    0.010196       0.904010  ...         0.585915  0.509317  0.581006
4  0.592452    0.009936       0.709599  ...         0.298311  0.362667  0.445690

[5 rows x 10 columns]

In [23]:
# 5 - 2 hidden layers each with 10 neurons
model_5 = MLPClassifier(hidden_layer_sizes=(10,10)).fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_5.hidden_layer_sizes)
print("n_layers_",model_5.n_layers_)

hidden layers sizes (10, 10)
n_layers_ 4


In [24]:
model_5_cv_results = pd.DataFrame(cross_validate(model_5,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_5_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  1.893991    0.008045       0.896845  ...         0.587805  0.530387  0.625162
1  0.213840    0.008476       0.895631  ...         0.497253  0.410959  0.500000
2  0.664982    0.007023       0.899149  ...         0.770833  0.265487  0.323851
3  0.400955    0.009696       0.908870  ...         0.685393  0.369748  0.452690
4  0.718617    0.009466       0.907655  ...         0.731707  0.441176  0.270880

[5 rows x 10 columns]

3 Describe after evaluating each model which model appears to have the best average f1 score by examining the results. (no need to compute the mean f1 simply a visual inspection is sufficient)

After evaluating each model it appears that model 4 (2 hidden layers each with 20 neurons) has the best average f1 score compared to the other models. Model 4 has the highest collective average and all of the f1 scores in the results are close/similar as well.

****Note after running this multiple times by doing a factory reset runtime and run all I noticed that the values change. At the time of this decision it appeared that 4 had the best average f1 score.

# Task 3

## Code Chunk 3 - Model 4 with Different Hyperparameters

In code chunk 3 the best model from step 2 that was determined is used to build two new models by changing the learning_rate_int, activation and the solver hyperparameters. After these two new models are made, the results are evaluated and explained in a text block on the attempt to improve the model by changing these hyperparameters.

In [25]:
# 1 Using the best model hidden layer sizes from Step 2 as you have determined the best f1 score:
  # Build 2 new models and change both the learning_rate_init, the activation, and the solver hyperparameters.

# model 4.1
model_4_1 = MLPClassifier(hidden_layer_sizes=(20,20), activation='logistic', learning_rate_init=0.002, solver='sgd').fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_4_1.hidden_layer_sizes)
print("n_layers_",model_4_1.n_layers_)

hidden layers sizes (20, 20)
n_layers_ 4


In [26]:
model_4_1_cv_results = pd.DataFrame(cross_validate(model_4_1,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_4_1_cv_results

fit_time  score_time  test_accuracy  ...  train_precision  test_f1  train_f1
0  0.512536    0.011135       0.890777  ...              0.0      0.0       0.0
1  1.066393    0.008277       0.889563  ...              0.0      0.0       0.0
2  1.671954    0.009262       0.890644  ...              0.0      0.0       0.0
3  0.332823    0.008734       0.890644  ...              0.0      0.0       0.0
4  1.474130    0.009396       0.890644  ...              0.0      0.0       0.0

[5 rows x 10 columns]

In [27]:
# model 4.2
model_4_2 = MLPClassifier(hidden_layer_sizes=(20,20), activation='tanh', learning_rate_init=0.015, solver='lbfgs').fit(CD_additional_modified_v2_enc,y_target)
print("hidden layers sizes",model_4_2.hidden_layer_sizes)
print("n_layers_",model_4_2.n_layers_)

hidden layers sizes (20, 20)
n_layers_ 4


In [28]:
model_4_2_cv_results = pd.DataFrame(cross_validate(model_4_2,
               CD_additional_modified_v2_enc,
               y_target,
               cv = 5,
               return_train_score=True,
               scoring=['accuracy','recall','precision','f1']))

model_4_2_cv_results

fit_time  score_time  test_accuracy  ...  train_precision   test_f1  train_f1
0  3.145734    0.009434       0.787621  ...         0.513433  0.454829  0.494253
1  3.070416    0.008201       0.888350  ...         1.000000  0.000000  0.011050
2  2.112494    0.008505       0.890644  ...         0.000000  0.000000  0.000000
3  0.078161    0.008173       0.890644  ...         0.000000  0.000000  0.000000
4  2.994188    0.008273       0.894289  ...         0.000000  0.155340  0.000000

[5 rows x 10 columns]

2 Evaluate the results of your new model. 
  Does it appear you have improved?
Explain in a text block your results of attempting to improve on the model.

From the results of the new models in task 3 based off the results of task 2, it does not appear that it has improved. My attempt to improve the model was to change some of the hyperparameters to hopefully get a positive effect and improvement. In the first rendition of changes I changed the activation parameter to 'logistic' and the solver to 'sgd as well as changing the learn rate to 0.002. The results came out as 0.0 for the f1 score so that was not good. It my second rendition I changed the activation to 'tanh', learning rate to 0.015 and solver to 'lbfgs'. I randomly adjusted these parameters and looking back I wish I would have only sligtly changed them to see which parameter had a greater affect in improving the results.

# Task 4

## Code Chunk 4 - GridSearchCV

Code chunk 4 is code that uses GridSearchCV to explore all of the hyperparameter combinations from all the existing models that were created in task 2 and task 3. Cross validation is set to 3 and f1 is used as the objective best score to attempt to maximize it. After running the GridSearchCV the results and some other questions/topics are explored and explained in text blocks.

In [29]:
# 1. using GridSearch explore ALL of the hyperparameter combination from all exiting models you have created
 # from both Step2 and Step3. Override GridSearchCV's default cv to be 3. Use f1 as your objective best score
  # to attempt to maximize.

parameters = {'learning_rate_init':[0.001,0.002,0.015],
               'hidden_layer_sizes':[(7,),(30,30),(25,10,10),(20,20),(10,10)],
              'activation':['logistic', 'tanh', 'relu'],
              'solver':['adam','sgd','lbfgs']
              }
mlp = MLPClassifier()
clf = GridSearchCV(mlp, parameters, scoring='f1', cv = 3)
clf.fit(CD_additional_modified_v2_enc, y_target)

clf.score(CD_additional_modified_v2_enc, y_target)

0.287215411558669

In [30]:
# Get the data out from the GridSearchCV Code and inspect it
grid_search_df = pd.DataFrame(clf.cv_results_)

print(grid_search_df.shape)

grid_search_df.sort_values('mean_test_score',ascending=False).head() 

(135, 15)


mean_fit_time  std_fit_time  ...  std_test_score  rank_test_score
90       0.736822      0.283791  ...        0.041363                1
96       0.330759      0.106214  ...        0.018882                2
30       0.897768      0.115009  ...        0.035293                3
27       1.411484      0.259977  ...        0.091038                4
11       3.168610      1.020372  ...        0.066809                5

[5 rows x 15 columns]

In [31]:
# best estimator the gridsearch discovered to get the best results
clf.best_estimator_

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(7,), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=None, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

1) In a text block explain how many hyperparameter combinations you will be evaluating and explain mathematically how you can be certain of this.

There are 135 hyperparameter combinations that will be evaluated. This is calculated by taking the number of learning_rate_init, hidden_layer_sizes, activation, and solver parameters and multiplying all of them. There are 3 learning rates, 5 hidden layer sizes, 3 activations, and 3 solvers. When you multiply them (3*5*3*5) you get 135.

2) In the same text block explain how many total models were evaluated when counting the cross-valildation.

Since each cross validation fold creates a model as does every hyperparameter combination you will take the value we get from the hyperparameter combinations (135) and times that by the cross validation fold size. In this case we are doing 3 fold cross validation and 135*3 equals 405 total models. 

# Task 5

## Code Chunk 5 - GridSeachCV exploring combinations of C and Kernel for the SVC Classifier

Code chunk 5 is similar to code chunk 4 but instead of ANN it uses the SVC classifier by exploring combinations of C and kernel. At least 20 hyperparameter combinations of C and kernal are explored for the SVC classifier. F1 score is still used as the objectevice to attempt to maximize. A text block is used to evaluate the results and compare to the ANN results.

In [32]:
# Using GridSearchCV explore at least 20 hyperparameter combinations of C and Kernel for the SVC classifier
# Use f1 as your objective best score to attempt to maximize
parameters = {'C':[5,10,15,20,30,33,40,50,60,70,80,90,100],
              'kernel':['rbf','poly','sigmoid']
              }
svc = SVC()
clf = GridSearchCV(svc, parameters,scoring='f1')
clf.fit(CD_additional_modified_v2_enc, y_target)

clf.score(CD_additional_modified_v2_enc, y_target)

0.40667678300455234

In [33]:
# best estimator discovered running the gridsearch in box 34
clf.best_estimator_

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='poly',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [34]:
# Print out the head of the resulting clf dataframe and sort it so that you can 
  # see the best mean score
grid_search_df = pd.DataFrame(clf.cv_results_)
print(grid_search_df.shape) 
grid_search_df.sort_values('mean_test_score',ascending=False).head()

(39, 15)


mean_fit_time  std_fit_time  ...  std_test_score  rank_test_score
37       0.249266      0.003147  ...        0.044758                1
34       0.255233      0.003553  ...        0.044441                2
31       0.249132      0.009227  ...        0.040291                3
28       0.258611      0.014617  ...        0.042953                4
25       0.246284      0.007384  ...        0.040787                5

[5 rows x 15 columns]

In a text block below enter this score value an explain how the cross validation and grid search works and how it arrived at this metric.

Explain if the model has exceeded the best ANN model we found. 

Of the two modeling options which appears to work the best?

The best mean f1 score from the code above is 0.413774	

Cross validation and grid search run through and make models based off of the hyperparmeters that were given in code block 34. As it does this, it finds the best estimator based on the objective we set, which in this case is the f1 score. The params that the cross validation and gridsearch got the best mean f1 score from was Cross function of 100 and a kernal of 'poly'. 

This model did not exceed the best ANN model we found. The best ANN model we found had a mean f1 score of 0.502979. So it was close but ANN was still higher.

The ANN model appears to work best given that it had a higher mean f1 score from the models we looked at. 

In [36]:
# Render A7_Gardner_Tyler.ipynb to an HTML output file.  
!cp "/content/drive/MyDrive/Colab Notebooks/A7_Gardner_Tyler.ipynb" ./

!jupyter nbconvert --to html "A7_Gardner_Tyler.ipynb"

[NbConvertApp] Converting notebook A7_Gardner_Tyler.ipynb to html
[NbConvertApp] Writing 367666 bytes to A7_Gardner_Tyler.html
